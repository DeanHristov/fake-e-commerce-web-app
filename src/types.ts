export enum PRODUCT_CATEGORIES {
  ELECTRONIC = 'ELECTRONIC',
}

export type TCurrency = 'USD' | 'BGN' | 'EUR';

export interface IComment {
  productId: string;
  _id: string;
  name: string;
  email: string;
  body: string;
  date: string;
}

// Object schema can be seen here: https://dummyjson.com/docs/products
export interface IProduct {
  id: number;
  title: string;
  description: string;
  price: number;
  discountPercentage: number;
  rating: number;
  stock: number;
  brand: string;
  category: string;
  thumbnail: string;
  images: string[];
}

export interface IUser {
  name: string;
  email: string;
  password: string;
  isAdmin: boolean;
}
