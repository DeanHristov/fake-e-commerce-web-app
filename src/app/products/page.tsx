import { FC } from 'react';
import { Metadata } from 'next';

export interface IProductsProps {}

export const metadata: Metadata = {
  title: 'Products',
};

const ProductsPage: FC<IProductsProps> = ({}) => {
  return <div className="h-[88vh]">Products page</div>;
};

export default ProductsPage;
