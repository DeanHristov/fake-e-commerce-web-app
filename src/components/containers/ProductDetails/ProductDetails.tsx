'use client';
import Image from 'next/image';
import { FC } from 'react';
import { IComment, TCurrency } from '../../../types';
import { Utils } from '../../../utils/Utils';
import Comment from '../../ui/Comment';
import Widget from '../Widget';

export interface IProductDetailsProps {
  id: string;
  name: string;
  image: string;
  currency: TCurrency;
  description: string;
  rating: number;
  price: number;
  comments?: IComment[];
}

const ProductDetails: FC<IProductDetailsProps> = ({
  id,
  name,
  image,
  description,
  rating,
  price,
  currency,
  comments = [],
}) => {
  return (
    <div className="flex-1">
      <Widget>
        <div className="flex flex-col lg:flex-row">
          <Image
            className="border-gray-300 border m-auto"
            alt=""
            width={440}
            height={320}
            src={image}
          />
          <div className="w-full">
            <h3 className="p-2  text-gray-500 border-b border-t border-gray-300">
              {name}
            </h3>
            <span className="block  text-gray-500 p-2 border-b border-gray-300">
              {Utils.parseAmountbyCurrency(price, currency)}
            </span>
            <span className="block p-2 border-b border-gray-300">{rating}</span>
          </div>
        </div>
        <p className="py-2 mb-10 text-gray-500">{description}</p>
        {comments.map((item: IComment) => (
          <Comment
            key={item._id}
            name={item.name}
            _id={item._id}
            body={item.body}
            date={item.date}
          />
        ))}
      </Widget>
    </div>
  );
};

export default ProductDetails;
