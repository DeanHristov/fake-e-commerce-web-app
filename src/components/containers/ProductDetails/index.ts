import ProductDetails, { IProductDetailsProps } from './ProductDetails';

export type { IProductDetailsProps };
export default ProductDetails;
