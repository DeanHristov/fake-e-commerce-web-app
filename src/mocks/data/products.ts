import { IProduct, PRODUCT_CATEGORIES } from '../../types';

const products: IProduct[] = [
  {
    _id: '29a34b48-8343-4e3d-8c21-3548ba249936',
    name: 'Airpods Pro2 Headphones',
    image: '/images/airpods.jpg',
    description:
      "Apple's new H2 chip is the driving force behind AirPods Pro and their improved sound characteristics. It works in unison with a personalized driver and amplifier for crystal clear heights and deep, rich bass with incredible clarity, so every sound is more pronounced than ever",
    brand: 'Apple',
    category: PRODUCT_CATEGORIES.ELECTRONIC,
    price: 399.99,
    countInStock: 10,
    rating: 4.5,
    currency: 'USD',
    reviews: 5,
  },
];

export default products;
