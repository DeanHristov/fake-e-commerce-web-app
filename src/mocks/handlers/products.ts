import { MockedRequest, rest, RestContext, RestHandler } from 'msw';
import mockedProducts from '../data/products';
import { IProduct } from '@/types';

export let productsHandler: RestHandler<MockedRequest<IProduct>>[] = [
  // Fetching the all products
  rest.get(`${process.env.API_URL}/products`, (req, res, ctx: RestContext) => {
    return res(
      ctx.status(200),
      ctx.json({
        products: mockedProducts,
        total: 100,
        skip: 0,
        limit: 30,
      }),
    );
  }),

  // Fetching a single product
  rest.get(
    `${process.env.API_URL}/products/:productId`,
    (req, res, ctx: RestContext) => {
      const { productId } = req.params;
      const outputItem: IProduct | undefined = mockedProducts.find(
        (item: IProduct) => item.id === Number(productId),
      );

      return res(ctx.status(200), ctx.json(outputItem ?? null));
    },
  ),
];
