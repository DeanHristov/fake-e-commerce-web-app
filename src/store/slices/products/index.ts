export * from './productsSlice';
export * from './selectors';
