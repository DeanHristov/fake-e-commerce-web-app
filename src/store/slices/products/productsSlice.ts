import { IProduct } from '@/types';
import { Utils } from '@/utils/Utils';
import { createSlice, PayloadAction } from '@reduxjs/toolkit';

const reducerName = 'products';
const initialState: Record<string, IProduct> = Utils.tryToLoadFromStorage(
  reducerName,
  {},
);

export const productsSlice = createSlice({
  name: reducerName,
  initialState,
  reducers: {
    initProducts: (
      state: Record<string, IProduct>,
      action: PayloadAction<IProduct[]>,
    ) => {
      for (let productItem of action.payload) {
        state[productItem._id] = productItem;
      }
    },
  },
});

export const { initProducts } = productsSlice.actions;
export default productsSlice;
