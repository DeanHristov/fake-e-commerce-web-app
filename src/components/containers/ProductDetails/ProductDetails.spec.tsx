import { render } from '@testing-library/react';
import ProductDetails from '.';
import { IProduct } from '@/types';
import products from '@/mocks/data/products';

const firstProduct: IProduct = products[0];

describe('Containers/ProductDetails <ProductDetails {...} />', () => {
  it('Should render a default state of the component', () => {
    const { container } = render(
      <ProductDetails
        price={firstProduct.price}
        currency={'EUR'}
        name={firstProduct.title}
        image={firstProduct.thumbnail}
        description={firstProduct.description}
        rating={firstProduct.rating}
      />,
    );

    expect(container).toMatchSnapshot();
  });
});
