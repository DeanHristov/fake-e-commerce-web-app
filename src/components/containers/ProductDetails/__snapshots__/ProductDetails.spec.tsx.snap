// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Containers/ProductDetails <ProductDetails {...} /> Should render a default state of the component 1`] = `
<div>
  <div
    class="flex-1"
  >
    <div
      class="bg-white border border-gray-300 rounded-lg p-4 shadow-lg"
    >
      <div
        class="py-0"
      />
      <div
        class="pt-2"
      >
        <div
          class="flex flex-col lg:flex-row"
        >
          <img
            alt=""
            class="border-gray-300 border m-auto"
            data-nimg="1"
            decoding="async"
            height="320"
            loading="lazy"
            src="/_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F1%2Fthumbnail.jpg&w=1080&q=75"
            srcset="/_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F1%2Fthumbnail.jpg&w=640&q=75 1x, /_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F1%2Fthumbnail.jpg&w=1080&q=75 2x"
            style="color: transparent;"
            width="440"
          />
          <div
            class="w-full"
          >
            <h3
              class="p-2  text-gray-500 border-b border-t border-gray-300"
            >
              iPhone 9
            </h3>
            <span
              class="block  text-gray-500 p-2 border-b border-gray-300"
            >
              â‚¬549.00
            </span>
            <span
              class="block p-2 border-b border-gray-300"
            >
              4.69
            </span>
          </div>
        </div>
        <p
          class="py-2 mb-10 text-gray-500"
        >
          An apple mobile which is nothing like apple
        </p>
      </div>
    </div>
  </div>
</div>
`;
